### ISSUE TITLE
Prepared statement with map - Codec not found for requested operation: [decimal <-> scala.math.BigDecimal]

### ISSUE THREAD

**Issue by mmatloka** [outworkers/phantom#411]
Columns `extends BigDecimalColumn(this)` work ok, but `extends MapColumn[SomeTable, Sth, Int, BigDecimal](this)` results in 
```
com.datastax.driver.core.exceptions.CodecNotFoundException: Codec not found for requested operation: [decimal <-> scala.math.BigDecimal]
	at com.datastax.driver.core.CodecRegistry.notFound(CodecRegistry.java:679) ~[cassandra-driver-core-3.0.0-rc1.jar:na]
	at com.datastax.driver.core.CodecRegistry.createCodec(CodecRegistry.java:540) ~[cassandra-driver-core-3.0.0-rc1.jar:na]
	at com.datastax.driver.core.CodecRegistry.findCodec(CodecRegistry.java:520) ~[cassandra-driver-core-3.0.0-rc1.jar:na]
	at com.datastax.driver.core.CodecRegistry.maybeCreateCodec(CodecRegistry.java:656) ~[cassandra-driver-core-3.0.0-rc1.jar:na]
	at com.datastax.driver.core.CodecRegistry.createCodec(CodecRegistry.java:538) ~[cassandra-driver-core-3.0.0-rc1.jar:na]
	at com.datastax.driver.core.CodecRegistry.findCodec(CodecRegistry.java:520) ~[cassandra-driver-core-3.0.0-rc1.jar:na]
	at com.datastax.driver.core.CodecRegistry.codecFor(CodecRegistry.java:470) ~[cassandra-driver-core-3.0.0-rc1.jar:na]
	at com.datastax.driver.core.AbstractGettableByIndexData.codecFor(AbstractGettableByIndexData.java:77) ~[cassandra-driver-core-3.0.0-rc1.jar:na]
	at com.datastax.driver.core.BoundStatement.bind(BoundStatement.java:196) ~[cassandra-driver-core-3.0.0-rc1.jar:na]
	at com.datastax.driver.core.DefaultPreparedStatement.bind(DefaultPreparedStatement.java:124) ~[cassandra-driver-core-3.0.0-rc1.jar:na]
	at com.websudos.phantom.builder.query.prepared.PreparedBlock.bind(PreparedBuilder.scala:169) ~[phantom-dsl_2.11-1.21.3.jar:1.21.3]
...
```

Phantom 1.21.3.

**Comment by mmatloka** [comment]
Workaround:

``` 
ContactPoints(hosts, port)
      .withClusterBuilder(_
        .withCodecRegistry(CodecRegistry.DEFAULT_INSTANCE.register(new BigDecimalCodec)))

...

class BigDecimalCodec extends MappingCodec(TypeCodec.decimal(), classOf[scala.math.BigDecimal]) {

  override def serialize(value: scala.math.BigDecimal): java.math.BigDecimal = value.bigDecimal

  override def deserialize(value: java.math.BigDecimal): scala.math.BigDecimal = BigDecimal(value)
}
```

**Comment by alexflav23** [comment]
Hi @mmatloka,

Yes, you have to go through the motions because of how poorly the Java driver addresses type safety. Not much we can do, but there is a mechanism so we are working on a default presets list to address some of this clutter.

Regards.

**Comment by mmatloka** [comment]
This workaround does not fully work.When I apply `column(row)` the runtime result returned contains java type which at some points gives me class cast exception between java and scala bigdecimal. May be related to `identity(obj).asInstanceOf[RR]` during primitives handling..

**Comment by mmatloka** [comment]
This occurs when I retrieve objects not via prepared statement, so the codec is not applied during read.

**Comment by alexflav23** [comment]
Hi @mmatloka,

I think we may need to add a `BigDecimalPrimitive` to the default set, that sounds like it would fix everything. Prepared statements probably need separate handling in the driver too, but the combo will fix things hopefully.

Regards.

**Comment by mmatloka** [comment]
Quick workaround `column(row).asInstanceOf[Map[Int, java.math.BigDecimal]].mapValues(BigDecimal(_))`

So finally it seems problem is that in cases of maps `BigDecimalPrimitive` is applied to `java.math.BigDecimal` so java value gets casted `identity(obj).asInstanceOf[RR]` to scala `BigDecimal` without proper conversion.

> I think we may need to add a BigDecimalPrimitive to the default set,

default set?

**Comment by mmatloka** [comment]
not yet fully working, need to add codec

**Comment by alexflav23** [comment]
Hi @mmatloka,

The default set meaning the default primitives that get imported into scope when you import:

`com.websudos.phantom.dsl._`.

So basically Maps use a pre-existing BigDecimal specialised primitive, I'd be surprised if that extraction fails. May I ask you for a full scenario(table definition + queries) so I can test on my own with that? I use big decimal columns a lot in several projects, as well as in maps, and I don't think I've hit this.

Regards.

**Comment by alexflav23** [comment]
Hi @mmatloka,

Just following up on this, apologies for the delay. I'm not entirely sure I can follow on what's wrong with it, are you referring to not being able to use this in a `MapColumn`? I remember we had similar problems with this for `DateTime`, and that's why we implemented the `PrimitiveType`:

```scala


  implicit object BigDecimalPrimitive extends Primitive[BigDecimal] {

    override type PrimitiveType = java.math.BigDecimal

    val cassandraType = CQLSyntax.Types.Decimal

    override def fromRow(column: String, row: Row): Try[BigDecimal] = nullCheck(column, row) {
      r => BigDecimal(r.getDecimal(column))
    }

    override def asCql(value: BigDecimal): String = value.toString()

    override def fromString(value: String): BigDecimal = BigDecimal(value)

    override def clz: Class[java.math.BigDecimal] = classOf[java.math.BigDecimal]
  }
```

So when this is parsed from a map, we look for `java.math.BigDecimal` and then parse things accordingly. From what I can understand you are talking about not being able to bind a `BigDecimal`, but otherwise the column behaves as expected, is that correct?

Regards.

**Comment by alexflav23** [comment]
Hi @mmatloka,

Just tested preparing a `BigDecimal`, seems like this works as expected. In the below case `Primitive` is simply a case class where every field is a phantom `Primitive` type. The `gen` will generate an instance of the `case class` with some arbitrary data.

```scala
val sample = gen[Primitive]

    val query = database.primitives.insert
      .p_value(_.pkey, ?)
      .p_value(_.long, ?)
      .p_value(_.boolean, ?)
      .p_value(_.bDecimal, ?)
      .p_value(_.double, ?)
      .p_value(_.float, ?)
      .p_value(_.inet, ?)
      .p_value(_.int, ?)
      .p_value(_.date, ?)
      .p_value(_.uuid, ?)
      .p_value(_.bi, ?)
      .prepare()

    val chain = for {
      store <- query.bind(sample).future()
      get <- database.primitives.select.where(_.pkey eqs sample.pkey).one()
    } yield get
```

**Comment by alexflav23** [comment]
Hi @mmatloka,

Found the problem. When we encountered this codec bit before in the driver, we improved `Primitive` to allow overriding how certain Java types get parsed and converted to their Scala counterparty, `BigDecimal` being such an example.

The workaround will be available in `1.28.7` today, I'm just adding the tests right now. Apologies for the delay on this, turns out it was an easy fix. `Primitive` defines a `PritimitiveType` and a `fromPrimitive` method, which defaults to `identity(obj).asInstanceOf[RR]`, but can be overriden on a per type basis, `BigDecimalPrimitive` now has the correct override for `BigDecimal.apply(javaBigDecimal)` instead of the direct type cast.
